src/collider/Geom.cpp

```diff
@@ -73,160 +68,161 @@ void Geom::CollideSphere(Sphere &sphere, void (*callback)(CollisionContact *)) c
 {
 	// PROFILE_SCOPED() // verbose profiling only, this gets called a LOT
-	if (maxContacts <= 0) return;
-	if (edgeNode->triIndicesStart) {
-		const GeomTree::Edge *edges = this->GetGeomTree()->GetEdges();
-		int numContacts = 0;
-		vector3f dir;
-		isect_t isect;
-		const std::vector<vector3f> &rVertices = GetGeomTree()->GetVertices();
-		for (int i = 0; i < edgeNode->numTris; i++) {
-			const int vtxNum = edges[edgeNode->triIndicesStart[i]].v1i;
-			const vector3d v1 = transToB * vector3d(rVertices[vtxNum]);
-			const vector3f _from(float(v1.x), float(v1.y), float(v1.z));
-
-			vector3d _dir(
-				double(edges[edgeNode->triIndicesStart[i]].dir.x),
-				double(edges[edgeNode->triIndicesStart[i]].dir.y),
-				double(edges[edgeNode->triIndicesStart[i]].dir.z));
-			_dir = transToB.ApplyRotationOnly(_dir);
-			dir = vector3f(&_dir.x);
-			isect.dist = edges[edgeNode->triIndicesStart[i]].len;
-			isect.triIdx = -1;
-
-			b->GetGeomTree()->TraceRay(btriNode, _from, dir, &isect);
-
-			if (isect.triIdx == -1) continue;
-			numContacts++;
-			const double depth = edges[edgeNode->triIndicesStart[i]].len - isect.dist;
-			// in world coords
-			CollisionContact contact;
-			contact.pos = b->GetTransform() * (v1 + vector3d(&dir.x) * double(isect.dist));
-			vector3f n = b->m_geomtree->GetTriNormal(isect.triIdx);
-			contact.normal = vector3d(n.x, n.y, n.z);
-			contact.normal = b->GetTransform().ApplyRotationOnly(contact.normal);
-			contact.distance = isect.dist;
-
-			contact.depth = depth;
-			contact.triIdx = isect.triIdx;
-			contact.userData1 = m_data;
-			contact.userData2 = b->m_data;
-			// contact geomFlag is bitwise OR of triangle's and edge's flags
-			contact.geomFlag = b->m_geomtree->GetTriFlag(isect.triIdx) |
-				edges[edgeNode->triIndicesStart[i]].triFlag;
-			callback(&contact);
-			if (--maxContacts <= 0) return;
-		}
-	} else {
-		CollideEdgesTris(maxContacts, edgeNode->kids[0], transToB, b, btriNode, callback);
-		CollideEdgesTris(maxContacts, edgeNode->kids[1], transToB, b, btriNode, callback);
+	const GeomTree::Edge *edges = GetGeomTree()->GetEdges();
+	const std::vector<vector3f> &rVertices = GetGeomTree()->GetVertices();
+
+	const GeomTree::Edge &edge = edges[edgeIdx];
+	const vector3d v1 = transToB * vector3d(rVertices[edge.v1i]);
+	const vector3f _from(float(v1.x), float(v1.y), float(v1.z));
+
+	vector3d dir = transToB.ApplyRotationOnly(vector3d(edge.dir));
+	isect_t isect;
+
+	isect.dist = edge.len;
+	isect.triIdx = -1;
+
+	isect_buf.clear();
+	b->GetGeomTree()->GetTriTree()->TraceRay(v1, 1.0 / dir, edge.len, isect_buf, triNode);
+
+	// TODO
+	// This is subtly dependent on overwriting intersections with the last triangle to be processed
+	// Ideally all contacts here would be reported...
+	vector3f traceDir = vector3f(dir);
+	for (uint32_t tri : isect_buf) {
```

Comment:
Why go through all the items here then? You can just take the last one.

Reply:
This is because the result from `GetTriTree()->TraceRay()` only returns intersections between the ray and the triangle's AABB. It does not guarantee an intersection between the ray and the triangle stored in the AABB leaf node. Thus, all potential contacts reported from the BVHTree need to be verified to determine if they're an actual contact against the underlying triangle data.